pkg/operation/botanist/component/nodelocaldns/nodelocaldns.go

```diff
@@ -0,0 +1,557 @@
+								VolumeMounts: []corev1.VolumeMount{
+									{
+										MountPath: "/run/xtables.lock",
+										Name:      "xtables-lock",
+										ReadOnly:  false,
+									},
+									{
+										MountPath: "/etc/coredns",
+										Name:      "config-volume",
+									},
+									{
+										MountPath: "/etc/kube-dns",
+										Name:      "kube-dns-config",
+									},
+								},
+							},
+						},
+						Volumes: []corev1.Volume{
+							{
+								Name: "xtables-lock",
+								VolumeSource: corev1.VolumeSource{
+									HostPath: &corev1.HostPathVolumeSource{
+										Path: "/run/xtables.lock",
+										Type: &hostPathFileOrCreate,
+									},
+								},
+							},
+							{
+								Name: "kube-dns-config",
+								VolumeSource: corev1.VolumeSource{
+									ConfigMap: &corev1.ConfigMapVolumeSource{
+										LocalObjectReference: corev1.LocalObjectReference{
+											Name: "kube-dns",
+										},
+										Optional: pointer.Bool(true),
+									},
+								},
+							},
+							{
+								Name: "config-volume",
+								VolumeSource: corev1.VolumeSource{
+									ConfigMap: &corev1.ConfigMapVolumeSource{
+										LocalObjectReference: corev1.LocalObjectReference{
+											Name: configMap.Name,
+										},
+										Items: []corev1.KeyToPath{
+											{
+												Key:  configDataKey,
+												Path: "Corefile.base",
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		vpa *autoscalingv1beta2.VerticalPodAutoscaler
+	)
+	utilruntime.Must(references.InjectAnnotations(daemonset))
+
+	if c.values.VPAEnabled {
+		vpaUpdateMode := autoscalingv1beta2.UpdateModeAuto
+		vpa = &autoscalingv1beta2.VerticalPodAutoscaler{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "node-local-dns",
+				Namespace: metav1.NamespaceSystem,
+			},
+			Spec: autoscalingv1beta2.VerticalPodAutoscalerSpec{
+				TargetRef: &autoscalingv1.CrossVersionObjectReference{
+					APIVersion: appsv1.SchemeGroupVersion.String(),
+					Kind:       "DaemonSet",
+					Name:       daemonset.Name,
+				},
+				UpdatePolicy: &autoscalingv1beta2.PodUpdatePolicy{
+					UpdateMode: &vpaUpdateMode,
+				},
+				ResourcePolicy: &autoscalingv1beta2.PodResourcePolicy{
+					ContainerPolicies: []autoscalingv1beta2.ContainerResourcePolicy{
+						{
+							ContainerName: autoscalingv1beta2.DefaultContainerResourcePolicy,
+							MinAllowed: corev1.ResourceList{
+								corev1.ResourceCPU:    resource.MustParse("50m"),
+								corev1.ResourceMemory: resource.MustParse("150Mi"),
+							},
+						},
+					},
+				},
+			},
+		}
+	}
+
+	return registry.AddAllAndSerialize(
+		serviceAccount,
+		podSecurityPolicy,
+		clusterRole,
+		roleBinding,
+		configMap,
+		service,
+		daemonset,
+		vpa,
+	)
+}
+
+func (c *nodeLocalDNS) bindIP() string {
+	if c.values.DNSServer != "" {
+		return nodeLocal + " " + c.values.DNSServer
+	}
+	return nodeLocal
+}
+func (c *nodeLocalDNS) containerArg() string {
```

Comment:
```suggestion

func (c *nodeLocalDNS) containerArg() string {
```

(same below, add some newline for readability)